Add documentation for 3d models in simulation (#10)
Team2122/2122_TeamTators: documentation/docs/Programming/3dModels.md

# 3D Models in AdvantageScope

## Table of Contents:
* [Documentation and Downloads](#other-documentation-and-downloads)
* [Where to get models](#where-to-get-models)
* [Exporting models from Onshape](#exporting-models-from-onshape)
* [Configuring and Exporting in CAD Assistant](#configuring-and-exporting-in-cad-assistant)
* [Combining Files in CAD Assistant](#combining-files-in-cad-assistant)
* [Moving Things Around in CAD Assistant](#moving-things-around-in-cad-assistant)
* [Model Folder and Files](#folder-and-files)
* [Naming Conventions and Model Descriptions](#naming-conventions-and-models)
* [Configuring Articulated Components](#configuring-articulated-components)
    * [Setting your Assets Folder](#setting-your-assets-folder)
    * [Configuring Base Positions](#configuring-base-positions)
* [Movement](#movement)
    * [Rotation Movement](#rotation---finding-the-point-of-rotation)
    * [Translation Movement](#translation---finding-the-starting-point)

> * [The Final Product](#the-final-product)

Reviewer: you don't need to generate a table of contents, the website will make one for you

## Other documentation and downloads

**It is extremely important to look at these documents, especially during your first creation of an articulated component!**

The following are other pieces of documentation needed to create models and articulate components in AdvantageScope. A few things are also mentioned here, but that doesn't mean you should skip these!

* [AdvantageScope Documentation](https://docs.advantagescope.org/more-features/custom-assets/)

This is the main resource besides this very file you are in now that explains implementing 3D models in AdvantageScope, made by team 6328.

**NOTE**: Do **NOT** just watch the video regarding 3D Robot models before creating your own model. Many things in the video are not explained, because you are supposed to **read the documentation** alongside it!

* [WPI-Lib Pose 3D](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/math/geometry/Pose3d.html)

The WPI-Lib documentation regarding the different methods used in the movement of articulated components.

* [CAD Assistant Download](https://dev.opencascade.org/project/cad-assistant)

> This program is the middleman in between Onshape and VS Code.

Reviewer: > middleman in between Onshape and VS Code

isn't it the middleman between onshape and advantage scope?

## Where to get models

The models for our robot and its components are found in our shared Onshape library.

> **NOTE**: In order to access this library, your account needs to be invited.

Reviewer 2: Can you give the reader a starting point for who to talk to / what to do to get an invitation?

> **NOTE**: In order to access this library, your account needs to be invited. In order to get invited, talk to Dylan Macmaster, who sends the invites via email.

Reviewer: Given that this is a publicly facing website, I'd like to avoid using full names. Maybe just "Dylan" or "the mentor who runs the CAD class in the fall" would work better

When you sign into Onshape, you should see this library if you were correctly invited and also accepted the invite:
![Sign in page with our frc library](../../docs\images\OnshapeLibrary.png)

## Exporting Models from Onshape

Exporting models is a fairly simple process.

First, click on the model (typically in a folder labelled that robot's year)

Go to the main assembly for the part you are trying to import, specifically one of the files with a **cube containing a shaded section**. 

**right click the file in the tab at the bottom of your screen**, and then press Export.

![ExportLocation](../../docs\images\OnshapeExport.png)

Export the file as a **STEP** file. You will need to make this .glb file in a second, but Onshape does not offer .glb exporting.

You do not really need to do anything else in the export menu apart from exporting to a .STEP file.

## Configuring and exporting in CAD Assistant

<!-- Note: This will need to be updated if we get a better CAD configuring and exporting software, as CAD assistant is...bad. Blender seems a bit complex for this task, but it could be an option. Preferably, a tool available across all OS would be best. -->

> Once you have your file exported from Onshape, you can't quite put it into VS Code yet. Instead, you must [download CAD Assistant]((https://dev.opencascade.org/project/cad-assistant)), a program that lets you export to .glb files.

Reviewer: wdym by "put it into VS Code?" VSCode does not handle 3d models

Author: You have to put the files into the code though?

Reviewer 3: yeah, but using the phrase "into VS Code" implies VS Code is the necessary arbiter between the cad and advantage scope, when it's just a tool to edit files on your computer that is entirely independent of that process

This program is also what we use for extra configuration of the models. (ie. Adding a non-moving component onto the [base model](#naming-conventions-and-models)). However, CAD Assistant is very limited in this aspect, making it hard to work with at times.

In most cases, you can simply open up your .STEP file, press the Save icon (floppy disk on left hand side), and press export as  a .glb file. At times (such as the example previously mentioned) though, you might need to move components around. 
![WhereSave](../../docs\images\WhereSaveButton.png)

> Location of the Save button

Reviewer: > Location of the Save button

This appears as regular text, but is better suited as an image caption. You can render it as such by using the `>` character. For example:

<pre>
> gud caption
</pre>

is rendered as

> gud caption

![WhereExport](../../docs\images\WhereExport.png)

> What to export as

Reviewer: > What to export as

see my other comment about image captions

> **NOTE**: if you did indeed read the AdvantageScope documentation closely, you might have seen a mention of glTF files. These are **not** used in our process, so please do not export into VS Code as a glTF. It will not work.

Reviewer: another mention of "export into VS Code," which doesn't really make sense

Reviewer: Additionally, could you add some explanation of why we don't use glTF files? The advantage scope docs explicitly say to use glTF files for all models

Author: Thinking about it now theres like no reason to NOT use glTF. I just didn't do it because the models being stored as STEPs works fine and it didnt seem neccesary to me to export from Onshape to the STEP model to glTF and then to glb. Plus, they said to follow the instructions to convert to glTF, but the instructions were to convert to glb, which was just confusing.

Author: They just use glTF for "storing the 3D models", not in the assets folder, which is what I was explaining, saying to not use glTF in the assets folder because it doesnt work

Reviewer 3: > "storing the 3D models"

if not storing them in the assets folder, is this just for storing before exporting to glb? do you know why the docs mention glTF at all if it can be avoided entirely and is unneeded? you may want to open up an issue in the advantage scope repository detailing your grievances if there's not already an issue about it

Author: I opened up the issue just now but who knows when theyll ever get back to me :(

Once exported, **the file will not be able to be viewed when clicked in the VS code editor!** Do not panic. This does not mean its broken, nor does it mean you should be using a glTF file instead. The VS code editor just cant process .glb files, since they are in binary.

### Combining files in CAD Assistant

> The base model for our robot should contain all parts that don't need to move separately from the robot (For example, the driverbase does not have any parts that move, while the picker does)

Reviewer 2: "For example, for our 2026 robot, the driverbase does not ..."

> The base model for our robot should contain all parts that don't need to move separately from the robot (For example, the 2026 robot model's driverbase does not have any parts that move, while the picker can rotate up and down)

Reviewer: ```suggestion
The base model for our robot should contain all parts that don't need to move separately from the robot (For example, the 2026 robot model's drivebase does not have any parts that move, while the picker can rotate up and down)
```

This often means that you need to combine other files into one another to put them on one model. To do this. Open up your current [model.glb](#naming-conventions-and-models).

> Then, select the folder icon on the left side and click the button labelled "Add to Current Document" on the bottom tab of the page. **You must do this BEFORE opening up the file you are trying to combine your base model with.**

Reviewer: can you add a picture for this?

Click on the file you are trying to combine, and it should appear on your base model. If it isn't where it should be, see the tutorial below.

### Moving things around in CAD Assistant

This can be done with a .STEP file just exported or, a .glb file currently in the code. 

> **CAD Assistant was not exactly designed for preforming this task**, but it is still possible. However, you should **try to avoid moving things around in CAD Assistant whenever you can**. One way to do this is to import the entire Robot Master and individually deleting non-static components from it, leaving only pieces that don't move.

Reviewer: ```suggestion
**CAD Assistant was not exactly designed for performing this task**, but it is still possible. However, you should **try to avoid moving things around in CAD Assistant whenever you can**. One way to do this is to import the entire Robot Master and individually deleting non-static components from it, leaving only pieces that don't move.
```

> **CAD Assistant was not exactly designed for preforming this task**, but it is still possible. However, you should **try to avoid moving things around in CAD Assistant whenever you can**. One way to do this is to import the entire Robot Master and individually deleting non-static components from it, leaving only pieces that don't move.

Reviewer: > try to avoid moving things around in CAD Assistant whenever you can

can you elaborate on this?

After all parts of the component you want to be moved are selected, click the icon **directly above the select parent icon**. Expand **location**, and then press **identity**.

![MoveComponent](../../docs\images\CADmove.png)

Arrows for the X, Y, and Z axis should appear in red, green and blue. You can use these arrows to position the part to where it needs to go.

Unfortunately, CAD Assistant is very finnicky about this part. You can position the component with these arrows, but it will not save this position. to get it to save the position, you must click a point in space to move the component to. 

On occasion, it will not follow the point that you clicked, and instead move to the bottom of your model. This makes it take a lot of patience to get it to get to the point you want it to at times. As previously stated, try to avoid moving things around in CAD Assistant when you can.

### **DO NOT FORGET TO SAVE YOUR CAD MODELS WHEN DONE!**

Saving your model is the same proccess as [Exporting](#configuring-and-exporting-in-cad-assistant) your file as a .glb for the first time. Of course, make sure it is still a .glb file when you save! 

## Folder and Files

> Next, **create a new folder** in VS Code. This is where your assets will be held. Inside that folder, create another new folder. **If you do not do this, then AdvantageScope cannot read the files**. You must name the nested folder "Robot_", and then whatever name you like. In your Robot_ file, it will have the config.json, as well as all your models. For details on setting up your config.json, watch the video on this page [here](https://docs.advantagescope.org/more-features/custom-assets/).

Reviewer: Can you go into more detail about the location and name of the first folder, maybe include an example from this year's codebase?

> Next, **create a new folder** in VS Code. This is where your assets will be held. Inside that folder, create another new folder. **If you do not do this, then AdvantageScope cannot read the files**. You must name the nested folder "Robot_", and then whatever name you like. In your Robot_ file, it will have the config.json, as well as all your models. For details on setting up your config.json, watch the video on this page [here](https://docs.advantagescope.org/more-features/custom-assets/).

Reviewer: ```suggestion
Next, **create a new folder** in VS Code. This is where your assets will be held. Inside that folder, create another new folder. **If you do not do this, then AdvantageScope cannot read the files**. You must name the nested folder "Robot_", and then whatever name you like. In your Robot_ folder, it will have the config.json, as well as all your models. For details on setting up your config.json, watch the video on this page [here](https://docs.advantagescope.org/more-features/custom-assets/).
```

### Naming Conventions and Models

> **A component should only be its own model if it moves separately from the robot**, such as a picker or a rotating turret. Each of your models should be named like as such:

Reviewer: ```suggestion
**A component should only be its own model if it moves separately from the robot**, such as a picker or a rotating turret. Each of your models should be named as such:
```

> **A component should only be its own model if it moves separately from the robot**, such as a picker or a rotating turret. Each of your models should be named like as such:

Reviewer: The vague description of "picker" does not imply a part that moves separately from the robot as I think you're intending, as an under the bumper picker that just has rollers and nothing else can still be interpreted as "moving separately from the robot" in that the rollers roll. Perhaps you could specify "rotating picker" or something along those lines?

> * **model.glb** - This is your base model, where every part that doesn't move separately goes. This includes things like the driverbase or where the game pieces are stored. If you need to add another part in order to update this, see [Combining Files in Cad Assistant](#combining-files-in-cad-assistant).

Reviewer: ```suggestion
* **model.glb** - This is your base model, where every part that doesn't move separately goes. This includes things like the drivebase or where the game pieces are stored. If you need to add another part in order to update this, see [Combining Files in Cad Assistant](#combining-files-in-cad-assistant).
```

> * **model.glb** - This is your base model, where every part that doesn't move separately goes. This includes things like the driverbase or where the game pieces are stored. If you need to add another part in order to update this, see [Combining Files in Cad Assistant](#combining-files-in-cad-assistant).

Reviewer: "where the gamepieces are stored" can often be several parts that all move, such as with the coral affector on last year's onseason robot

* **model0.glb** - This is your first separately moving component. From 0, the number increases (ie: model1.glb, model2.glb...)

**NOTE**:  In your config.json, the componenets must be in the same order as your models. If a component's file is named model1.glb, it cannot be first in the array.

To keep track of your different models, create variables in the config.json that signify the model number and name, such as seen in the components example below.

![Components Example](../../docs\images\ComponentsExample.png)

## Configuring Articulated Components

### Setting your Assets folder
Simulate your code and open up AdvantageScope. Click "App", and then "Use Custom Assets Folder".

> Copy the path of the folder that holds your [Robot_](#folder-and-files) file, and set that as your assets folder.

Reviewer: ```suggestion

Copy the path of the folder that holds your [Robot_](#folder-and-files) folder, and set that as your assets folder.
```

### Configuring base positions

> Now, we get to the most important part of the process: putting your models into action. For now, put the following log into your code:

Reviewer: Can you specify where this goes in code?

Logger.recordOutput("Zeroed Component Poses", new Pose 3d[] {new Pose 3d()});

> 

Reviewer: You can mark code in markdown using backticks. For example:

\```
console.log("foo");
// super useful comment on another line
\```
and \`console.log("bar"); // inline`

are rendered as
```
console.log("foo");
// super useful comment on another line
```
and `console.log("bar"); // inline`

What this does:

* Creates what is called a pose 3d, which, as the name suggests, is the current pose of the object in a 3d space

* "Zeroed" in this case means that it will be based off of your config.json file positions

> 

Reviewer: The way you've written this sort of implies advantage scope uses the positions in config.json because the word "Zeroed" is in the log key

Click the + at the top right, and then 3d field. In the bottom right corner, set the field to Axes and zoom in on the robot model. (Right now, it is likely a KitBot. We will change this in just a second)

> 

Reviewer: > Click the + at the top right

we have changed scenery to advantage scope from our IDE but this has not been noted

Author: what?

Drag in the DriverPose log, clicking the arrow icon and changing the model name to the name of your robot that you put in the config.json previously. Below the DriverPose log, drag your Zeroed Component Poses in. If the icon is green and says "Vision Target", click on it and change to a component.

> 

Reviewer: > Drag in the DriverPose log

1. "log" references an entire file full of logged data. I think you're looking for the word "key" here, as in "key/value pair"
2. "DriverPose" is not a key that exists. I think you can just leave it as simply "Drag in the robot pose"

> 

Reviewer: > Below the DriverPose log, drag your Zeroed Component Poses in.

Can you add a picture of this so as to distinguish between dragging the value underneath of the robot pose and in the poses field as a whole

**Drag one of these in for every component you are trying to add to your 3D model**, and make sure they are where they should be on the robot. If they aren't, go back and tweak your config.json file.

> ![HowNestComponents](../../docs\images\NestedRobotPose.png)

Reviewer: What is the importance of this image?

Author: It ssupposed to show how the component is under the Robot pose and not separate from it

While a few things are different name-wise because this is from the [AdvantageScope documentation](https://docs.advantagescope.org/more-features/custom-assets/), the above picture shows an example of that your Zeroed Components will look like in logging before being configured. (**Remember! The AdvantageScope Documentation contains very important info not covered here!**)

> 

Reviewer: ```suggestion
While a few things are different name-wise because this is from the [AdvantageScope documentation](https://docs.advantagescope.org/more-features/custom-assets/), the above picture shows an example of what your zeroed components will look like in logging before being configured. (**Remember! The AdvantageScope Documentation contains very important info not covered here!**)

```

> 

Reviewer: > While a few things are different name-wise

different in what sense? What kind of things are you talking about?

### Movement
This portion requires a lot of information from the WPI methods section, linked [here](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/math/geometry/Pose3d.html)

> 

Reviewer: > from the WPI methods section

It's not really the "WPI methods section," but the javadocs page for `Pose3d`. The former implies the page goes over every method for every class in WPILib, while the ladder implies that it's a page going over all of the methods and constructors available for the `Pose3d` class specifically

Once your models have been configured using Zeroed Component Poses, the next step is to make them move.

In the log() of the subsystem of the component you want to move, you need to do the following things: 

> 

Reviewer: ```suggestion
In the `log()` method of the subsystem of the component you want to move, you need to do the following things: 

```

> * Make sure that the Robot is being simulated, so the code doesn't create a bunch of things it doesn't need while the robot is actually running

Reviewer: > Make sure that the Robot is being simulated

can you give a code example of what this means

> * In the Constants file of your subsystem, create a Translation 3D named something along the lines of SUBSYSTEM_STARTING_POINT and set it to 0,0,0. You will adjust this in a minute.

Reviewer: ```suggestion
* In the constants file of your subsystem, create a Translation 3D named something along the lines of `SUBSYSTEM_STARTING_POINT` and set it to (0,0,0). You will adjust this in a minute.
```

Then, depending on if your component rotates about a point or moves up/down or left/right, you do the following things:

#### Rotation - Finding the point of Rotation

* Create a double called timer and set it to timer.getTimeStamp()
    * This creates a constantly increasing double
* Create a rotation 3D and set it to time
    * this will make the component rotate in a full circle, which helps to pinpoint the point of rotation

> * Create a Pose 3d and set it to k.Zero, then make it rotate around (your subsystem's point of rotation, the value rotation 3d you just made) (for more context about .rotateAround(), see the WPI documentation)

Reviewer: ```suggestion
* Create a `Rotation3d` and set it to time
    * this will make the component rotate in a full circle, which helps to pinpoint the point of rotation
* Create a `Pose3d` and set it to `Pose3d.kZero`, then make it rotate around your subsystem's point of rotation, which is the `Rotation3d` you just made (for more context about .rotateAround(), see the WPI documentation)
```

> * Create a Pose 3d and set it to k.Zero, then make it rotate around (your subsystem's point of rotation, the value rotation 3d you just made) (for more context about .rotateAround(), see the WPI documentation)

Reviewer: > for more context about .rotateAround(), see the WPI documentation

can you link to the docs for this specific methodi on the javadocs page

>     * Log said Pose 3d

Reviewer: could you give a code example of this?

>     * Log said Pose 3d

Reviewer: ```suggestion
    * Log said `Pose3d`
```

Before you continue to find the point of rotation, make sure that your component is rotating about the part of the component you want it to. If it looks like it is rotating about a different part of the component, set your rotation 3d to 0 and configure.

> 

Reviewer: > Before you continue to find the point of rotation, make sure that your component is rotating about the part of the component you want it to

isn't that exactly what continuing to find the point of rotation is

* Once your component is rotating in a circular motion, use trial and error to move your component to the point on the robot you want it to rotate about. Start by making big changes (such as changing a coordinate value by 0.1) to your starting point, and as you get closer to the point, make smaller and smaller adjustments.

> 

Reviewer: ```suggestion
* Once your component is rotating in a circular motion, use trial and error to move your component to the point on the robot you want it to rotate about. Start by making big changes to your starting point, (such as changing a coordinate value by 0.1) and as you get closer to the point, make smaller and smaller adjustments.

```

![RotationPointEx](../../\docs\images\RotationPoint.png)

> this is an example of a rotation point. Notice both positive and negative values, as well as the fact that these numbers are pretty small. This is because the measurements are in meters, relative to the bottom center of the robot.

Reviewer: ```suggestion
This is an example of a rotation point. Notice both positive and negative values, as well as the fact that these numbers are pretty small. This is because the measurements are in meters, relative to the bottom center of the robot.
```

#### Rotation - Making it move as it does in real life

* Now, set your rotation 3d so that one of the values (roll, pitch, or yaw) is set to inputs.currentPosition * 2pi.

>     * It is multiplied by 2pi because that is the equation to convert rotations to radians 

Reviewer: > set your rotation 3d so that one of the values (roll, pitch, or yaw) is set to inputs.currentPosition * 2pi

this assumes that `inputs.currentPosition` is a double in units of rotations that exists, which may not always be true. I would just say something along the lines of "set the value of the `Rotation3d` to the angle of your subsystem, and ensure the angle is in units of radians."

![Rotation Example](../../docs\images\RotationEx.png)

#### Translation - Finding the starting point

* Create a Translation 3d, setting the X, Y, and Z values to your startingPoint.getX(), startingPoint.getY(), and startingPoint.getZ() respectively. 

* Use trial and error to find the starting point, beginning with large adjustments in the values and slowly refining

>     * values for the X, Y, and Z coordinates are in meters relative to the robot, so a "big" adjustment would be considered to be 0.1 or 0.2 in either a negative or positive direction.

Reviewer: Shouldn't we adjust `SUBSYSTEM_STARTING_POINT` directly?

Author: you are adjusting it directly

#### Translation - Making it move as it does in real life
* Simply add inputs.currentPosition to whichever one of the startingPoint.get X(), Y(), or Z() values
    * this will often make it move in either direction far too much. To solve this, use trial and error to find a coefficient to inputs.currentPosition that will make it move less.

> 

Reviewer: This isn't necessarily sage advice for subsystems that have canonical units, which is most subsystems. This is really only applicable for subsystems that don't have canonical units, like the vertical extension from this year

    ![Translation Example](../../docs\images\TranslationEx.png)

> 

Reviewer: I think this would fit better as two dropdowns for translation and rotation. See this page for details on how to do that: https://blog.markdowntools.com/posts/markdown-collapsible-sections-guide

#### Testing
* Bind the subsystem moving to a new desired position to a button on the driver controller and test it out in simulation!

>     * use "driverController.button().onTrue(...);"

Reviewer: ```suggestion
    * use `driverController.button().onTrue(...);`
```

    as opposed to whileTrue, and use different buttons to set it to different positions for best testing

## The Final Product

Hopefully, you should have a 3d Simulation of the robot!
Here is an example of what it should look like, with parts able to move up and down when you make the desired position of their subsystems change.

>  ![Final Bot](../../docs\images\EpicRobot.png)

Reviewer: Can you make this a gif of the robot parts moving?
